Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,355 @@
+-/
+instance : Equivalence TuringEquivalent :=
+  {
+    refl := TuringEquivalent.refl,
+    symm := TuringEquivalent.symm,
+    trans := @TuringEquivalent.trans
+  }
+
+/--
+Instance declaring that `RecursiveIn` is a preorder.
+-/
+instance : IsPreorder (ℕ →. ℕ) RecursiveIn where
+  refl := RecursiveIn.refl
+  trans := @RecursiveIn.trans
+
+/--
+The Turing degrees as the set of equivalence classes under Turing equivalence.
+-/
+def TuringDegree :=
+  Antisymmetrization _ RecursiveIn
+
+/--
+The `join` function combines two partial functions `f` and `g` into a single partial function.
+For a given input `n`:
+
+- **If `n` is even**:
+  - It checks if `f` is defined at `n / 2`.
+  - If so, `join f g` is defined at `n` with the value `2 * f(n / 2)`.
+
+- **If `n` is odd**:
+  - It checks if `g` is defined at `n / 2`.
+  - If so, `join f g` is defined at `n` with the value `2 * g(n / 2) + 1`.
+-/
+def join (f g : ℕ →. ℕ) : ℕ →. ℕ :=
+  fun n =>
+    if n % 2 = 0 then
+      (f (n / 2)).map (fun x => 2 * x)
+    else
+      (g (n / 2)).map (fun y => 2 * y + 1)
+
+/-- Join notation `f ⊕ g` is the "join" of partial functions `f` and `g`. -/
+infix:99 "⊕" => join
+
+/--
+For any partial functions `a`, `b₁`, and `b₂`, if `b₁` is Turing equivalent to `b₂`,
+then `a` is Turing reducible to `b₁` if and only if `a` is Turing reducible to `b₂`.
+-/
+lemma reduce_lifts₁ : ∀ (a b₁ b₂ : ℕ →. ℕ), b₁≡ᵀb₂ → (RecursiveIn a b₁) = (RecursiveIn a b₂) := by
+  intros a b₁ b₂ bEqb
+  apply propext
+  constructor
+  · intro aRedb₁
+    apply RecursiveIn.trans aRedb₁ bEqb.1
+  · intro aRedb₂
+    apply RecursiveIn.trans aRedb₂ bEqb.2
+
+/--
+For any partial functions `f`, `g`, and `h`, if `f` is Turing equivalent to `g`,
+then `f` is Turing reducible to `h` if and only if `g` is Turing reducible to `h`.
+-/
+lemma reduce_lifts₂ : ∀ (f g h : ℕ →. ℕ),
+f ≡ᵀ g → (RecursiveIn f h = RecursiveIn g h) := by
+  intros f g h fEqg
+  apply propext
+  constructor
+  · intro fRedh
+    apply RecursiveIn.trans fEqg.2 fRedh
+  · intro gRedh
+    apply RecursiveIn.trans fEqg.1 gRedh
+
+/--
+Here we show how to lift the Turing reducibility relation from
+partial functions to their Turing degrees, using the above lemmas.
+-/
+def TuringDegree.turing_red (d₁ d₂ : TuringDegree) : Prop :=
+  @Quot.lift₂ _ _ Prop (TuringEquivalent)
+  (TuringEquivalent) (RecursiveIn) (reduce_lifts₁) (reduce_lifts₂) d₁ d₂
+
+/--
+Instance declaring that `TuringDegree.turing_red` is a partial order.
+-/
+instance : PartialOrder TuringDegree where
+  le := TuringDegree.turing_red
+  le_refl := by
+    apply Quot.ind
+    intro a
+    apply RecursiveIn.refl
+  le_trans := by
+    apply Quot.ind
+    intro a
+    apply Quot.ind
+    intro b
+    apply Quot.ind
+    intro c
+    exact RecursiveIn.trans
+  le_antisymm := by
+    apply Quot.ind
+    intro a
+    apply Quot.ind
+    intro b
+    intros aRedb bReda
+    apply Quot.sound
+    constructor
+    · exact aRedb
+    · exact bReda
```

Comment:
All of this can be rewritten to use existing `Antisymmetrization` stuff, like
```lean
instance : PartialOrder TuringDegree :=
  @instPartialOrderAntisymmetrization (ℕ →. ℕ)
    {le := RecursiveIn, le_refl := .refl, le_trans _ _ _ := RecursiveIn.trans}
```